plugins/c8y_firmware_plugin/src/firmware_manager.rs

```diff
@@ -0,0 +1,197 @@
+use crate::child_device::get_child_id_from_child_topic;
+use crate::child_device::FirmwareOperationResponse;
+use crate::common::mark_pending_firmware_operation_failed;
+use crate::common::ActiveOperationState;
+use crate::download::FirmwareDownloadManager;
+use c8y_api::http_proxy::C8YHttpProxy;
+use c8y_api::smartrest::message::collect_smartrest_messages;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestFirmwareRequest;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestRequestGeneric;
+use c8y_api::smartrest::topic::C8yTopic;
+use mqtt_channel::Connection;
+use mqtt_channel::Message;
+use mqtt_channel::MqttError;
+use mqtt_channel::SinkExt;
+use mqtt_channel::StreamExt;
+use mqtt_channel::TopicFilter;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+use tedge_api::health::health_check_topics;
+use tedge_api::health::health_status_down_message;
+use tedge_api::health::send_health_status;
+use tokio::sync::Mutex;
+use tracing::error;
+use tracing::info;
+
+pub const PLUGIN_SERVICE_NAME: &str = "c8y-firmware-plugin";
+const FIRMWARE_UPDATE_RESPONSE_TOPICS: &str = "tedge/+/commands/res/firmware_update";
+
+pub struct FirmwareManager {
+    mqtt_client: Connection,
+    c8y_request_topics: TopicFilter,
+    health_check_topics: TopicFilter,
+    firmware_update_response_topics: TopicFilter,
+    firmware_download_manager: FirmwareDownloadManager,
+}
+
+impl FirmwareManager {
+    pub async fn new(
+        tedge_device_id: String,
+        mqtt_port: u16,
+        http_client: Arc<Mutex<dyn C8YHttpProxy>>,
+        local_http_host: String,
+        tmp_dir: PathBuf,
+        timeout_sec: Duration,
+    ) -> Result<Self, anyhow::Error> {
+        let mqtt_client = Self::create_mqtt_client(mqtt_port).await?;
+
+        let c8y_request_topics = C8yTopic::SmartRestRequest.into();
+        let health_check_topics = health_check_topics(PLUGIN_SERVICE_NAME);
+        let firmware_update_response_topics =
+            TopicFilter::new_unchecked(FIRMWARE_UPDATE_RESPONSE_TOPICS);
+
+        let firmware_download_manager = FirmwareDownloadManager::new(
```

Comment:
Although I understand that this design was inspired by the `c8y-config-plugin`, a separate `FirmwareDownloadManager` abstraction is less useful/relevant here as this plugin doesn't have a "upload" notion which would required another `FirmwareUploadManager`. So, it'd be simpler to keep everything in the `FirmwareDownloadManager` in this `FirmwareManager` struct itself, to avoid an unnecessary level of indirection.

Reply:
We removed the abstraction of `FirmwareDownloadManager` together :)